File: it-agents/skills/code-architect/references/code-examples/ruby/factory_method.rb
# frozen_string_literal: true

# Factory Method — subclasses decide which product class to instantiate via a factory method.

# Product role: anything that can render itself.
class TextExport
  def serialize(payload) = "TEXT|#{payload}"
end

class JsonExport
  def serialize(payload) = %({"value":"#{payload}"})
end

# Creator declares the factory method and uses the product it returns.
class Reporter
  # Factory method — overridden by subclasses.
  def build_formatter
    raise NotImplementedError, "#{self.class} must define #build_formatter"
  end

  def report(payload)
    formatter = build_formatter
    "[#{self.class.name}] #{formatter.serialize(payload)}"
  end
end

class TextReporter < Reporter
  def build_formatter = TextExport.new
end

class JsonReporter < Reporter
  def build_formatter = JsonExport.new
end

if __FILE__ == $PROGRAM_NAME
  [TextReporter.new, JsonReporter.new].each do |reporter|
    puts reporter.report("hello")
  end
end

File: it-agents/skills/code-architect/references/code-examples/swift/FactoryMethod.swift
// Factory Method — subclasses decide which product class to instantiate via a factory method.

protocol Transport {
    func deliver() -> String
}

struct GroundUnit: Transport {
    func deliver() -> String { "carrying cargo by road" }
}

struct AirUnit: Transport {
    func deliver() -> String { "carrying cargo by air" }
}

// Creator declares the factory method; subclasses override it to choose a product.
class Dispatcher {
    func makeTransport() -> Transport {
        fatalError("subclass must override makeTransport()")
    }

    func runRoute() -> String {
        let transport = makeTransport()
        return "Dispatcher is " + transport.deliver()
    }
}

class RoadDispatcher: Dispatcher {
    override func makeTransport() -> Transport { GroundUnit() }
}

class SkyDispatcher: Dispatcher {
    override func makeTransport() -> Transport { AirUnit() }
}

func runDemo() {
    let dispatchers: [Dispatcher] = [RoadDispatcher(), SkyDispatcher()]
    for dispatcher in dispatchers {
        print(dispatcher.runRoute())
    }
}

runDemo()
